include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_ASYNC_POOL_H_
+#define DALI_CORE_MM_ASYNC_POOL_H_
+
+#include <algorithm>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+#include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/free_list.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/cuda_event_pool.h"
+#include "dali/core/cuda_stream.h"
+
+namespace dali {
+namespace mm {
+
+template <memory_kind kind, class FreeList, class LockType, class Upstream = memory_resource<kind>>
+class async_pool_base : public stream_aware_memory_resource<kind> {
+ public:
+  explicit async_pool_base(Upstream *upstream) : global_pool_(upstream) {
+  }
+  ~async_pool_base() {
+    try {
+      synchronize();
+    } catch (const CUDAError &e) {
+      if ((e.is_rt_api() && e.rt_error() != cudaErrorCudartUnloading) ||
+          (e.is_drv_api() && e.drv_error() != CUDA_ERROR_DEINITIALIZED))
+        std::terminate();
+    }
+  }
+
+  /**
+   * @brief Waits until all pending frees are finished.
+   */
+  void synchronize() {
+    {
+      std::lock_guard<LockType> guard(lock_);
+      for (auto &kv : stream_free_) {
+        if (!kv.second.free_list.head)
+          continue;
+        if (!sync_stream_)
+          sync_stream_ = CUDAStream::Create(true);
```

Comment:
Why not create it in the constructor?

Reply:
If there's nothing to clean, we shouldn't even create this stream.

Reply:
What is the cost of the creating stream?
What is the cost of having one created ahead?
Maybe we can create it to have the `synchronize()` execution time more predictable?

Reply:
It would bind the constructor to a particular device that was set as current during construction. Having that said, perhaps I should store the current device at 1st allocation and restore it in the destructor with some `deviceguard`.

Reply:
Added DeviceGuard.